README
Module-Which version 0.0205
===========================

'Module::Which' is the basis of the script 'which_pm' intented
to show which version of a Perl module is installed (if it is there at all).

Modules are searched by name (like 'YAML') or by subcategories
('DBD::' means all modules under the DBD subdirectories of
your Perl installation, matching both 'DBD::Oracle' and 'DBD::ODBC::Changes').

This module is very simple and most won't need it.
But it has been instructive for the author to see how many broken modules
one can find under your Perl installation (some which don't accept
even a 'require' statement), modules with no version number and
documentation files (named '.pm') which do not return a true value.

To find out modules under subcategories, L<Module::Find> by Christian
Renz was used.

Well, all that said, this module is no more than automating:

  perl -MInteresting::Module -e 'print $Interesting::Module'

or better the one-liner

  perl -e '$pm = shift; eval "require $pm"; print ${"${pm}::VERSION"}' DBI

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  GetOpt::Long
  Pod::Usage
  Test::More
  File::Spec::Unix
  File::Glob
  File::Find
  ExtUtils::MakeMaker

and recommends

  Test::Pod

TO DO

* needs a more comprehensive test suite.

* implement more options on which() behavior

  return => 'ARRAY' | 'HASH' | 'HASH(FIRST)' | 'HASH(MULT)' | 'HASH(LIST)'
  verbose => 0 | 1

  include => @INC
  method => 'REQUIRE' | 'EXTRACT'
  info   => @fields


* finish documentation of Data::Hash::Transform
* stuff Data::Hash::Transform into its own distribution

* fix bug in --help option
* introduce more options to control C<which_pm> behavior: C<--include>, 
  C<--method>, etc.

* always update "META.yml" before releasing (this is not
  being done automatically because MakeMaker overwrite
  keys like 'license' and 'recommends')

* which_pm should produce nice output for narrow terminal window sizes.
* which_pm should be able to report the version of perl (and other Perl scripts)
* a more complete implementation of --version for which_pm 
  (copyright and license included)

* support the pattern '' as a shortcut to show all installed modules
  in include directories
* finish documentation of Module::Which::List

COPYRIGHT AND LICENCE

Copyright (C) 2005 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.




Changes
Revision history for Perl module Module::Which

0.04 2014-06-26 NEILB
    - bin/which_pm didn't have a $VERSION, which was causing a warning.
      It now uses the $VERSION from Module::Which. Thanks to FERREIRA :-)

0.03 2014-06-24 NEILB
    - Fixed the SYNOPSIS to show working code, and the two types of result
      data you can get from which(). RT#16332
    - Clarify in the doc that the return data includes the path to the module.
    - Added github repo to doc.
    - Switched to Dist::Zilla
    - Reformatted this file as per CPAN::Changes::Spec
    - Removed Module::Downgrade, as only aiming to support 5.6+

0.0207 2008-03-05 FERREIRA
    - mention Module::Finder in SEE ALSO
    - recommends Term::Size::Any instead of Term::Size
    - hacked up support for "which_pm perl"
    - TODO: document that "which_pm perl" works

0.0206 2007-11-20 FERREIRA
    - no code change
    - fixed packaging issues

0.0205 2006-05-21 FERREIRA
    - BUG FIX: in CPAN RT #16332, Jeff Tanner 
      noticed that the synopsis code didn't work. 
      Indeed, &which was using "return => 'ARRAY'" 
      as default instead of "return => 'HASH'"

0.0204 2006-05-21 FERREIRA
    - Term::Size is recommended
    - when determining terminal size, non-tty STDOUT
      prevents truncation
    - if possible, terminal width is retrieved with
      the help of Term::Size

0.02_03 2005-12-27 FERREIRA
    - new switch -I for specifying the library paths
    - correspoding changes to Module::Which
    - new switch -w for specifying terminal width

0.02_02 2005-12-08 FERREIRA
    - module versions are obtained via MM->parse_file
      after using EE::MM

0.02_01 2005-10-28 FERREIRA
    - now which_pm uses String::Truncate when file paths
      are too long (instead of simple truncate)
    - documented bug about wrong versions when multiple
      instances are found in library path

0.02 2005-10-10 FERREIRA
    - just the same as 0.01_05
    - it's a shame: the only purpose of this
      release it to raise my kwalitee to 16.0
      as development versions aren't counted

0.01_05 2005-07-27 FERREIRA
    - added a "--p5p" option to which_pm so that Config-related
      paths can be shown
    - wrote tests for the previously introduced modules
    - wrote Module::Which::P5Path to make Config-related
      translations: like "/usr/lib/perl5/Module/Which.pm" to
      '${sitelib}/Module/Which.pm'

0.01_04 2005-06-22 FERREIRA
    - file path was include in the information about
      a Perl module 'which_pm' gives
    - replaced "t/Module-Which.t" with "t/001basic.t"
      which uses the mock modules under "t/tlib" to test
    - the bug of not being able to find multiple installed
      versions of a module with an exact pattern like 'YAML'
      is gone with the use of Module::Which::List
    - subcategories are now written DBD:: instead of DBD::*
    - replaced Module::Find dependency by a new module
      Module::Which::List which allows me to recover
      the module path
    - option --version reports a sensible program name with
      "basename($0)"
    - added --version option to which_pm with corresponding docs 

0.01_03 2005-06-21 FERREIRA
    - improved a bit the output of 'which_pm' by finding
      out the maximum length of a module name before printing
    - temporarily, "META.yml" will not be automatically
      updated by MakeMaker (so that the correct license is set)
    - "bin/which_pm.PL" renamed to "bin/which_pm"
    - added "t/003pod.t" to test POD for errors 
      (when Test::Pod 1.00 is there)
    - completed documentation of Data::Hash::Transform
      which was truncated and buggy

0.01_02 2005-06-16 FERREIRA
    - options '--verbose' and '--quiet' supported by now
    - use of Getopt::Long and Pod::Usage in which_pm.pl
    - bug in Module::Which documented
    - test script for Data::Hash::Transform: "t/002hash.t"
    - code for converting from an array of hashes to hash of hashes
      moved to Data::Hash::Transform

0.01_01 2005-06-15 FERREIRA
    - experimental code to downgrade the source
      code to 5.005
    - which_pm is not verbose by default anymore:
      set environment variable WHICH_PM_VERBOSE 
    - options to customize which() behavior:
      'return' and 'verbose'
    - which() now returns an array ref by default
      and optionally a hash ref
    - added reference to existing CPAN modules 

0.01 2005-06-14 FERREIRA
    - first release to CPAN
    - created by h2xs 1.23 with options -AX -n Module::Which



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "find version and path of locally installed modules",
  "AUTHOR" => "Adriano R. Ferreira <ferreira\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Module-Which",
  "EXE_FILES" => [
    "bin/which_pm"
  ],
  "LICENSE" => "perl",
  "NAME" => "Module::Which",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Find" => 0,
    "File::Glob" => 0,
    "File::Spec" => 0,
    "File::Spec::Unix" => 0,
    "Getopt::Long" => 0,
    "Pod::Usage" => 0,
    "String::Truncate" => 0,
    "Term::Size::Any" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "0.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Find" => 0,
  "File::Glob" => 0,
  "File::Spec" => 0,
  "File::Spec::Unix" => 0,
  "Getopt::Long" => 0,
  "Pod::Usage" => 0,
  "String::Truncate" => 0,
  "Term::Size::Any" => 0,
  "Test::More" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





inc/Module/Downgrade.pm
package Module::Downgrade;

use 5.005;
use strict;
#use warnings unless $] < 5.006;

require Exporter;

#no ours
use vars qw(@ISA @EXPORT $VERSION);
@ISA = qw(Exporter);
@EXPORT = qw(downgrade);

$VERSION = '0.00_02';

# make substitutions to Perl 5.6.1 code to make it (hopefully)
# work with Perl 5.005
sub s_pm {
	my $f = shift;
	open SOURCE, "<$f"
		or die "can't open '$f': $!";
	open OUT, ">$f.out"
		or die "can't open '$f.out': $!";
	while (<SOURCE>) {
		if (/^\s*#/) {
			print OUT; next;
		}
		if (s/use\s+5.006/#$&/) {
			print OUT;
			print OUT "use 5.005;\n";
			next;
		}
		if (s/use\s+warnings/#$&/) {
			print OUT;
			next;
		}
		if (s/our\s+([\$@%]\w+)(.*?);/#$&/) {
			print OUT;
			print OUT "use vars qw($1);\n";
			print OUT "$1$2;\n";
			next;
		}
		print OUT;
	}
	close OUT;
	close SOURCE;
	rename $f, "$f.orig"
		or die "can't rename '$f' to '$f.orig': $!";
	rename "$f.out", $f
		or die "can't rename '$f.out' to '$f': $!";

}

sub downgrade {
	for my $pm (@_) {
		s_pm($pm);
	}
}

1;
__END__

=head1 NAME

Module::Downgrade - Tries to turn Perl 5.6 code to Perl 5.005

=head1 SYNOPSIS

  use Module::Downgrade;
  downgrade('Mod.pm', 'script.pl');

=head1 DESCRIPTION

C<Module::Downgrade> is a joke for trying to be compatible.
It depends entirely on the generosity of the code being translated.

Don't use it. It is not ready for general use.

=head2 EXPORT

C<downgrade> is exported by default.

=head1 SEE ALSO

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Which>.

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut


lib/Module/Which/List.pm

package Module::Which::List;
$Module::Which::List::VERSION = '0.04';
use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);
our @EXPORT_OK = qw(list_pm_files);

use File::Glob qw(bsd_glob);
#use File::Spec::Functions qw(abs2rel);
require File::Spec::Unix;
use File::Find qw(find);

# returns a list with no duplicates
sub uniq {
    my %h;
    return grep { ! $h{$_}++ } @_;
}


# my @files = _plain_list_files('File/*', @INC);
sub _plain_list_files {
    my $glob = shift;
    my @INC = @_;
    my @files;
    #push @files, bsd_glob("$_/$glob") for @INC;
    for my $base (@INC) {
        #print "bsd_glob($_/$glob)\n";
        push @files, 
             map { { path => $_, rpath => File::Spec::Unix->abs2rel($_, $base), base => $base } }
             grep { -e } bsd_glob("$base/$glob")
    }

    return @files if wantarray;
    return \@files;
}

# my @files = _deep_list_files('Data/**.pm', @INC)
sub _deep_list_files {
    my $glob = shift;
    my @INC = @_;

    #print "deep_list_files: ...\n";

    my $root = $glob;
    $root =~ s/\*\*\.pm$//;

    my $base;
    my @files;

    my $wanted = sub {
        if (/\.pm$/) {
            push @files, { path => $_, rpath => File::Spec::Unix->abs2rel($_, $base), base => $base };
        }
    };

    for (@INC) {
        $base = $_;
        if (-e "$base/$root" && -d "$base/$root") {
            find({ wanted => $wanted, no_chdir => 1 }, "$base/$root");
        }
    }

    return @files if wantarray;
    return \@files;

}

sub list_files {
    my $glob = shift;
    my %options = @_;
    my @include;
    @include = ($options{include}) ? @{$options{include}} : @INC;
    @include = uniq @include 
        if exists $options{uniq} ? $options{uniq} : 1; # FIXME: need documentation !!!!
    #print "include: @include\n";
    my @files;
    if ($glob =~ /\*\*\.pm$/) {
        return _deep_list_files($glob, @include);
    } else {
        return _plain_list_files($glob, @include);
    }
}

sub file_to_pm {
    my $rpath = shift;
    $rpath =~ s|\.pm$||;
    $rpath =~ s|/|::|g;
    return $rpath;
}

sub list_pm_files {
    my $pm_glob = shift;
    my %options = @_;

    my $glob = $pm_glob;
    $glob =~ s|::$|::*| unless $options{recurse};
    $glob =~ s|::$|::**| if $options{recurse};
    $glob =~ s|::|/|g;
    $glob =~ s|$|.pm|;
    #print "glob: $glob\n";

    my @pm_files = list_files($glob, @_);
    
    my @pm;
    for (@pm_files) {
        push @pm, { 
             pm => file_to_pm($_->{rpath}), 
             path => $_->{path}, 
             base => $_->{base} 
        };
    }

    return @pm if wantarray;
    return \@pm;

}

1;

__END__

=head1 NAME

Module::Which::List - Lists C<.pm> files under specified library paths

=head1 SYNOPSIS

  use Module::Which::List qw(list_pm_files);

  my @files = list_pm_files('XML::', 'lib1/', 'lib2/');
  # return all modules XML::* installed under 'lib1/' and 'lib2/'

  my @files = list_pm_files('Data::Dumper'); # uses @INC

=head1 DESCRITION

Yes, I know: it is a mess down below. But release early, release often
(before my breath disappears).


    pm_glob => 'File::Which'
    pm_root => 'File::Which::'


    list_files($pm_glob, { recurse => '0|1', include => \@INC })

        prefixes - take a look at Module::List
        pod      - take a look at Module::List

    list_files('Module::Which')

    => [ { pm => 'Module::Which', path => '/usr/lib/perl5/site_perl/5.8/Module/Which.pm' } ]

    list_files('Module::')

    => [ 
        { pm => 'Module::Build', path => '/usr/lib/perl5/site_perl/5.8/Module/Build.pm' } 
        { pm => 'Module::Find', path => '/usr/lib/perl5/site_perl/5.8/Module/Find.pm' }, 
        { pm => 'Module::Which', path => '/usr/lib/perl5/site_perl/5.8/Module/Which.pm' } 
       ]

=head1 BUGS

This documentation, in such a status, is a bug.

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Which>.

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut


lib/Data/Hash/Transform.pm

package Data::Hash::Transform;
$Data::Hash::Transform::VERSION = '0.04';
use 5.006;
use strict;
use warnings;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(hash_f hash_l hash_m hash_a hash_em);

use Carp qw(croak);

=head1 NAME

Data::Hash::Transform - Turns array of hashes to hash of hashes in predefined ways

=head1 SYNOPSIS

  use Data::Hash::Transform qw(hash_f hash_l hash_m hash_a hash_em);

  my $loh = [ { k => 1, n => 'one' }, { k => 2, n => 'two' }, { k => 1, n => 'ein' } ];
  $hoh1 = hash_f($loh, 'k'); # keep first
  $hoh2 = hash_l($loh, 'k'); # keep last
  $hoh3 = hash_m($loh, 'k'); # keep a list (if needed)
  $hoh4 = hash_a($loh, 'k'); # always keep a list

  $hoh = hash_em($loh, 'k', $meth); # $meth is one of 'f', 'l', 'm', or 'a'

=head1 DESCRIPTION

This module provides four algorithms to turn an array of hashes 
to a hash of hashes. The transformation is based on using
the value at a certain key of inner hashes as the key 
in the outer hash.

So:

  [ { k => 1, n => 'one' }, { k => 2, n => 'two' } ]

turns to

  { 1 => { k => 1, n => 'one' }, 2 => { k => 2, n => 'two } }

when C<'k'> is the key of keys. (From this example, it was
made obvious that here we mean array and hash refs when talking about
arrays and hashes.)

The difference among the algorithms happen when the same key happens
twice or more. For example, how do the following array maps
to a hash? (C<'k'> is still the key of keys here.)

  [ { k => 1, n => 'one' }, { k => 2, n => 'two' }, { k => 1, n => 'ein' } ]

The following alternatives (among others) are possible:

=over 4

=item *

keep the first

  { 1 => { k => 1, n => 'one' }, 2 => { k => 2, n => 'two' } }

=item *

keep the last

  { 2 => { k => 2, n => 'two' }, 1 => { k => 1, n => 'ein' }  }

=item *

keep a list in the case of collisions

  { 1 => [ { k => 1, n => 'one' }, { k => 1, n => 'ein' } ],
    2 => { k => 2, n => 'two' } }

=item *

always keep a list (for the case of collisions)

  { 1 => [ { k => 1, n => 'one' }, { k => 1, n => 'ein' } ],
    2 => [ { k => 2, n => 'two' } ] }

=back

That is exactly what we implement here.

=head2 EXPORT

None by default. C<hash_f>, C<hash_l>, C<hash_m>, C<hash_a>,
C<hash_em> can be exported on demand.

=cut


# keep last (remember (l)ast)
sub hash_l {
	my ($ary, $kk) = @_;
	my %hash;
	$hash{$_->{$kk}} = $_ for @$ary;
	return \%hash;
}

# note. The implementation takes for granted that
# the inner hashes have $kk as keys. If they don't
# C<undef> will turn to C<''> and things can get
# messed up.

# keep first (remember (f)irst)
sub hash_f {
	my ($ary, $kk) = @_;
	my %hash;
	for (@$ary) {
		my $k = $_->{$kk};
		$hash{$k} = $_ unless exists $hash{$k};
	}
	return \%hash;
}

# keep an array in case of collisions (remember (m)ulti)
sub hash_m {
	my ($ary, $kk) = @_;
	my %hash;
	for (@$ary) {
		my $k = $_->{$kk};
		if (exists $hash{$k}) {
			$hash{$k} = [ $hash{$k} ] if ref $hash{$k} ne 'ARRAY';
			push @{$hash{$k}}, $_;
		} else {
			$hash{$k} = $_;
		}
	}
	return \%hash;
}

# always keep an array (remember (a)rray)
sub hash_a {
	my ($ary, $kk) = @_;
	my %hash;
	for (@$ary) {
		my $k = $_->{$kk};
		if (exists $hash{$k}) {
			push @{$hash{$k}}, $_;
		} else {
			$hash{$k} = [ $_ ];
		}
	}
	return \%hash;
}

# all of them together
sub hash_em {
	my ($ary, $kk, $m) = @_;
	my %methods = ( l => \&hash_l, f => \&hash_f, m => \&hash_m, a => \&hash_a );
	my $method = $methods{$m || 'f'}
		or croak "hash_em method '$m' unknown: should be one of 'l', 'f', 'm', or 'a'";
	return &$method($ary, $kk);
}

=pod

=head1 HASH_M VERSUS HASH_A

The difference between using C<hash_m> and C<hash_a> is
primarily oriented to the code that is going to consume
the transformed hash. In the case of C<hash_m>, it must
be ready to handle two cases: a single element which appears
as a hash ref and multiple elements which appear as an
array ref of hash refs. In the case of C<hash_a>, 
the treatment is more homogeneous and you will always
get an array ref of hash refs.

A typical code with the return of C<hash_m> is illustrated
by the code below.

  my $h = hash_m($loh);
  while (my ($k, $v) = each %$h) {
	  if (ref $v eq 'ARRAY') {
		  do something with $_ for @$v;
	  } else {
		  do something with $v
	  }
  }

or the shorter:

  my $h = hash_m($loh);
  while (my ($k, $v) = each %$h) {
	  my @vs = (ref $v eq 'ARRAY') ? @$v : ($v);
	  do something with $_ for @vs;
  }

With C<hash_a>, it would look like:

  my $h = hash_m($loh);
  while (my ($k, $v) = each %$h) {
	  do something with $_ for @$v;
  }

It is a trade-off: the client code can be simple (C<hash_a>)
or the overhead of data structures can be reduced (C<hash_m>).

=head1 TO DO

If you are familiar with L<XML::Simple>, you probably have
recognized some of the tranformations it does with hashes against arrays.
Mainly, the ones represented by C<hash_m> and C<hash_l>
(when C<ForceArray> is used).

Other transformations based on typical behavior of
L<XML::Simple> are possible. For example,

=over 4

=item *

discard the key element

  [ { k => 1, n => 'one' }, { k => 2, n => 'two' } ]

to

  { 1 => { n => 'one' }, 2 => { n => 'two' } }

and even (for C<'n'> defined to be the contents key)

  { 1 => 'one', 2 => 'two' }

=item *

mark the key element

  [ { k => 1, n => 'one' }, { k => 2, n => 'two' }, { k => 1, n => 'ein' } ]

to

  { 1 => { -k => 1, n => 'one' }, 2 => { -k => 2, n => 'two' } }

=back

Maybe someday this gets implemented too. 

=head1 ISSUES

The functions C<hash_*> have been designed to be fast
and that's why their code is redundant. One could write a 
function with all bells and whistles which does all the
work of them together, by using options and querying them
at runtime. I think the code would be slightly harder to maintain
and perfomance may suffer. But this is just guessing.
Soon I will write such an implementation and a benchmark
to make sure it is worth to use this code as it is.

=for comment
=head1 SEE ALSO

=head1 BUGS

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Which>.

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut

1;



lib/Module/Which.pm
package Module::Which;
$Module::Which::VERSION = '0.04';
use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA = qw( Exporter );
our @EXPORT = qw( which );

#use Module::Find;
use Module::Which::List qw(list_pm_files);
use Data::Hash::Transform qw(hash_em);

#sub pm_require {
#    my $pm = shift;
#    my $verbose = shift;
#    eval "require $pm";
#    if ($@) { # error
#        warn "'require $pm' failed: $@" if $verbose;
#        return (0, $@);
#    }
#    return 1
#}

#sub pm_version {
#    my $pm = shift;
#    no strict 'refs';
#    return ${"${pm}::VERSION"};
#}

use ExtUtils::MakeMaker;

=begin private

=item B<pm_version>

  $v = pm_version($pm);

Parses a PM file and return what it thinks is $VERSION
in this file. (Actually implemented with 
C<use ExtUtils::MakeMaker; MM->parse_version($file)>.)
C<$pm> is the filename (eg., F<lib/Data/Dumper.pm>).

=end private

=cut
sub pm_version {
    my $pm = shift;
    my $v;
    eval { $v = MM->parse_version($pm); };
    return $@ ? undef : $v;
}

sub pm_info {
    my $pm = shift; # --- { pm: (pm), path: (path), base: (base), }
    my $options = shift;

    my $version = pm_version($pm->{path});

    return { 
        version => $version, 
        pm      => $pm->{pm}, 
        path    => $pm->{path}, 
        base    => $pm->{base}  
    };
}

#sub is_wildcard {
#    return shift =~ /::\*$/;
#}

#sub expand_wildcard {
#    my $wildcard = shift;
#    $wildcard =~ s/::\*$//;
#    return findallmod $wildcard;
#}

# turns an array of hashes to a hash of hashes
sub hashify (\@$) {
    my ($ary, $opt_meth) = @_;
    our %meth = ( 'HASH' => 'f', 'HASH(FIRST)' => 'f', 'HASH(MULTI)' => 'm', 'HASH(LIST)' => 'a' );
    my $meth = $meth{$opt_meth}
        or die "hash strategy '$opt_meth' unknown";
    return hash_em($ary, 'pm', $meth);
}


# which(@pm)
# which(@pm, $options) where $options is a hash ref
sub which {
    my $options = {};
    $options = pop @_ if ref $_[-1];
    $options->{return} = 'HASH' unless $options->{return};

    my @pm = @_;

    my @info;
    for my $pm (@pm) {

        # special case: 'perl'
        if ( $pm eq 'perl' ) { 
            push @info, { 
                pm => 'perl', 
                version => sprintf("%vd", $^V), #$],
                path => $^X,
                base => '', # XXX ?!
             };
             next;
        }

        #push @info, pm_info($_, $options) for list_pm_files($pm, recurse => 1);

        my @pm_files = list_pm_files($pm, recurse => 1, include => $options->{include});
        if (@pm_files) {
            push @info, pm_info($_, $options) for @pm_files;
        } else {
            push @info, { pm => $pm };
        }

        #if (is_wildcard($pm)) {
        #    push @info, pm_info($_, $options) for expand_wildcard($pm);
        #} else {
        #    push @info, pm_info($pm, $options);
        #}
    }
    return \@info if $options->{return} eq 'ARRAY';

    return hashify(@info, $options->{return});
}

1;

__END__

=head1 NAME

Module::Which - find version and path of locally installed modules

=head1 SYNOPSIS

  use Module::Which qw/ which /;

  my $result = which('Module::Which', 'YAML', 'XML::', 'DBI', 'DBD::');
  while (my ($module, $info) = each %$result) {
      print "$module:\n":
      print "  version: $info->{version}\n" if $info->{version};
      print "     path: $info->{path}\n"    if $info->{path}; 
  }

Or you can request an array ref instead of a hash ref:

  my $result = which('strict', 'YAML', {return => 'ARRAY'});

  foreach my $info (@$result) {
      print "$info->{pm}:\n":
      print "  version: $info->{version}\n" if $info->{version};
      print "     path: $info->{path}\n"    if $info->{path}; 
  }

=head1 DESCRIPTION

C<Module::Which> provides the C<which()> function, which takes
the name of one or more modules, and returns information about
those modules if they're intalled locally, including the version
and the path.

C<Module::Which> is the basis of the script C<which_pm> which
displays the retrieved information to STDOUT.

Modules are searched by name (like 'YAML') or by subcategories
('DBD::' means all modules under the DBD subdirectories of
your Perl installation, matching both 'DBD::Oracle' and 'DBD::ODBC::Changes').

This module is very simple and most won't need it.
But it has been instructive for the author to see how many broken modules
one can find under your Perl installation (some which don't accept
even a 'require' statement), modules with no version number and
documentation files (named '.pm') which do not return a true value.

=for comment
To find out modules under subcategories, L<Module::Find> by Christian
Renz was used.

Well, all that said, this module is no more than automating:

  perl -MInteresting::Module -e 'print $Interesting::Module::VERSION'

or better the one-liner

  perl -e '$pm = shift; eval "require $pm"; print ${"${pm}::VERSION"}' DBI

=over 4

=item B<which>

  my $info = which(@pm)
  my $info = which(@pm, { return => 'ARRAY', verbose => 1 }

Returns an array ref with information about the modules specified 
(by name or '::*' patterns). This information is a hash ref which
actually contains:

=over 4

=item * pm: the name of the Perl module

=item * version: the installed version

=item * path: the full path to the Perl module

=back

The version is the one found by accessing the scalar variable C<$VERSION> 
of the package, after a I<require> statement.
If the module was not found, 'version' is C<undef>. If the
module has no C<$VERSION>, 'version' is C<'undef'> (the string).
If the 'require' statement failed, 'version' is 'unknown'.

A hash ref of options can be given as the last argument.
The option C<return> can take one of the values: 'ARRAY', 'HASH',
'HASH(FIRST)', 'HASH(MULTI)', 'HASH(LIST)'.
'HASH' is the default and means to return a hash ref.
'HASH' forces the return of a hash ref where the module name
is used as key. 'ARRAY' is used to get an array ref.

The different strategies for returning a hash are different 
only if the same module is found twice or more times
in the current search path. 'HASH' which is the same as 
'HASH(FIRST)' only considers the first occurrence.
'HASH(MULTI)' will store multiple values in an array ref
(if needed). The problem with MULTI is that sometimes
you get a hash ref and sometimes an array ref of hash refs.
If 'HASH(LIST)' is used, an array ref will be stored
always, even if there is only one occurrence.

The option C<verbose> can be set to turn on and off
warnings on requiring the sought modules.

=back

=head2 EXPORT

C<which> is exported by default.

=head1 SEE ALSO

L<Module::Find> was my friend to implement this module as a breeze.
But I have found some itches and wrote my own L<Module::Which::List>
based on this and L<Module::List> by Andrew Main. 

After releasing it into CPAN, I found

    Module::InstalledVersion
    Module::Info
    Module::List
    Module::Locate
    Module::Finder

Module::InstalledVersion has a different approach (it does not run 
the modules to find
out their versions, but extract them via regexes) and does not
has a command-line interface which was the main thrust of this
distribution. I have been studying the others too.

=head1 BUGS

Known bugs:

=over 4

=item *

When a module is found twice or more in the library path,
the version is the one of the first file.

=back

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Which>.

=head1 REPOSITORY

L<https://github.com/neilbowers/Module-Which>

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut
